torch_xla/core/xla_model.py

```diff
@@ -615,19 +620,83 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
             f"`output` needs to be a list of Tensors, but given {type(output)}."
         )
       if len(output) != len(value):
-        raise ValueError("`output` length doesn't match `input` length: "
-                         f"{len(output)} vs {len(input)}.")
-      # Call the out of place version of the reduce_scatter
-      new_token = torch_xla._XLAC._xla_all_gather_coalesced_out(
-          output, value, token, dim, shard_count, groups or [], pin_layout)
-      torch_xla._XLAC._set_all_reduce_token(devctx.device, new_token)
-      return output
-
-    result = torch_xla._XLAC._xla_all_gather_coalesced(value, token, dim,
-                                                       shard_count, groups or
-                                                       [], pin_layout)
-    torch_xla._XLAC._set_all_reduce_token(devctx.device, result[-1])
-    return result[:-1]
+        raise ValueError("`output` length doesn't match `value` length: "
+                         f"{len(output)} vs {len(value)}.")
+
+    # helper function for bucketing
+    def _all_gather_coalesced(tensor_list, output_list=[]):
+      if output_list != []:
+        if len(output_list) != len(tensor_list):
+          raise ValueError(
+              "_all_gather_coalesced: `output_list` length doesn't match `tensor_list` length: "
+              f"{len(output_list)} vs {len(tensor_list)}.")
+        # Call the out of place version of the reduce_scatter
+        new_token = torch_xla._XLAC._xla_all_gather_coalesced_out(
+            output_list, tensor_list, token, dim, shard_count, groups or [],
+            pin_layout)
+        torch_xla._XLAC._set_all_reduce_token(devctx.device, new_token)
+        return output_list
+
+      result = torch_xla._XLAC._xla_all_gather_coalesced(
+          tensor_list, token, dim, shard_count, groups or [], pin_layout)
+      torch_xla._XLAC._set_all_reduce_token(devctx.device, result[-1])
+      return result[:-1]
+
+    #return _all_gather_coalesced(value, output if output else [])
+
+    total = 0
+    tensor_bucket = []
+    output_bucket = []
+    out_tensors = []
+    bucket_cap = int(
+        os.getenv("ALL_GATHER_REDUCE_SCATTER_BUCKET_CAP_MB",
+                  _ALL_GATHER_REDUCE_SCATTER_BUCKET_CAP_MB)) * 1024 * 1024
+    if groups:
+      divisor = len(groups[0]) if type(groups[0]) == list else len(groups)
+    else:
+      divisor = xrt_world_size()
+    bucket_cap = bucket_cap / divisor
+
+    for idx, tensor in enumerate(value):
+      tensor_bytes = tensor.numel() * tensor.element_size()
+
+      # Tensor is larger than bucket_cap, don't bucketize
+      if tensor_bytes > bucket_cap:
```

Comment:
Similar to https://github.com/pytorch/xla/pull/6417/files#r1511919328 ,  maybe this if section can be removed to simplify.